### PR TITLE
feat: Initial support for multiple @PrimaryKey annotations.

### PR DESCRIPTION
**Pull request by bdferris-v2**
As discussed in #1187 and https://bit.ly/gtfs-validator-multi-column-primary-keys

- [x] Run the unit tests with `gradle test` to make sure you didn't break anything
- [x] Format the title like "feat: [new feature short description]". Title must follow the Conventional Commit Specification(https://www.conventionalcommits.org/en/v1.0.0/).
- [x] Linked all relevant issues

### CODE REVIEW

File: processor/src/main/java/org/mobilitydata/gtfsvalidator/processor/GtfsFileDescriptor.java [hunk 1]
```diff
@@ -49,11 +48,19 @@ public GtfsFieldDescriptor getFieldByName(String name) {
...
-  public abstract Optional<GtfsFieldDescriptor> primaryKey();
+  public abstract ImmutableList<GtfsFieldDescriptor> primaryKeys();
 
-  public abstract Optional<GtfsFieldDescriptor> firstKey();
+  public boolean hasSingleColumnPrimaryKey() {
+    return primaryKeys().size() == 1;
+  }
+
+  public GtfsFieldDescriptor getSingleColumnPrimaryKey() {
+    return primaryKeys().get(0);
```
[the remark this answers is not in the document]
Comment (bdferris-v2) [reply processor/src/main/java/org/mobilitydata/gtfsvalidator/processor/GtfsFileDescriptor.java:60]: Done.

File: processor/src/main/java/org/mobilitydata/gtfsvalidator/processor/EndRangeValidatorGenerator.java [hunk 2]
```diff
@@ -147,8 +147,8 @@ private static CodeBlock generateNoticeContext(
     TypeName tableLoaderTypeName = new GtfsEntityClasses(fileDescriptor).tableLoaderTypeName();
     CodeBlock.Builder block =
         CodeBlock.builder().add("$T.FILENAME, entity.csvRowNumber(), ", tableLoaderTypeName);
-    if (fileDescriptor.primaryKey().isPresent()) {
-      block.add("entity.$L(), ", fileDescriptor.primaryKey().get().name());
+    if (fileDescriptor.hasSingleColumnPrimaryKey()) {
+      block.add("entity.$L(), ", fileDescriptor.getSingleColumnPrimaryKey().name());
```
[the remark this answers is not in the document]
Comment (bdferris-v2) [reply processor/src/main/java/org/mobilitydata/gtfsvalidator/processor/EndRangeValidatorGenerator.java:151]: I think there may be some other notices that assume a single-column primary key.  Fixing something like this would likely require a refactor of the individual notices, but I suppose that would be a breaking change for users who assume the notice schema is fixed?  @maximearmstrong do you have any thoughts on refactoring notice schemas?

File: core/src/main/java/org/mobilitydata/gtfsvalidator/table/GtfsTableContainer.java [hunk 3]
```diff
@@ -66,15 +66,9 @@ public boolean hasColumn(String columnName) {
...
-   *
-   * <p>The first returned key (if present) is annotated with {@link
-   * org.mobilitydata.gtfsvalidator.annotation.PrimaryKey} or {@link
-   * org.mobilitydata.gtfsvalidator.annotation.FirstKey}.
-   *
-   * <p>The second returned key (if present) is annotated with {@link
-   * org.mobilitydata.gtfsvalidator.annotation.SequenceKey}.
+   * <p>A table may have zero, one, or multiple primary keys, so the returned list contains from 0
+   * to N items. Keys are columns annotated with {@link
+   * org.mobilitydata.gtfsvalidator.annotation.PrimaryKey}.
```
[the remark this answers is not in the document]
Comment (maximearmstrong) [reply core/src/main/java/org/mobilitydata/gtfsvalidator/table/GtfsTableContainer.java:70]: What would be the pros of having a fixed order? Leaving order unspecified is fine with me, but if you have a strong opinion on this @asvechnikov2, let us know.

File: processor/src/main/java/org/mobilitydata/gtfsvalidator/processor/EndRangeValidatorGenerator.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (maximearmstrong) [reply processor/src/main/java/org/mobilitydata/gtfsvalidator/processor/EndRangeValidatorGenerator.java:151]: Refactoring the notice schemas would be a breaking change indeed. Let's avoid it if we can.

Though, I'm not sure to understand why single column primary key entities could be a problem. Could you elaborate on this?

File: processor/src/main/java/org/mobilitydata/gtfsvalidator/processor/TableContainerIndexGenerator.java [hunk 4]
```diff
@@ -0,0 +1,403 @@
...
+            .addParameter(TypeName.get(indexField.javaType()), "key")
+            .returns(ParameterizedTypeName.get(ClassName.get(Optional.class), entityTypeName))
+            .addStatement("return Optional.ofNullable($L.getOrDefault(key, null))", fieldName)
+            .build());
+  }
+
+  private static void addMapByCompositeKey(
+      TypeSpec.Builder typeSpec, ImmutableList<GtfsFieldDescriptor> keys, TypeName entityTypeName) {
+
+    // Field: Map<CompsiteKey, EntityType?> byCompositeKeyMap;
```
Comment (aababilov) [inline processor/src/main/java/org/mobilitydata/gtfsvalidator/processor/TableContainerIndexGenerator.java:180]: Typo: CompositeKey

File: core/src/main/java/org/mobilitydata/gtfsvalidator/table/GtfsTableContainerWithMultiColumnPrimaryKey.java [hunk 5]
```diff
@@ -0,0 +1,27 @@
...
+ *
+ * @param <T> the GTFS entity type for this container
+ */
+public interface GtfsTableContainerWithMultiColumnPrimaryKey<T, K extends MulitColumnKey> {
+
+  /**
+   * Marker interface for all composite key implementation classes used by multi-column primary key
+   * table containers.
+   */
+  interface MulitColumnKey {}
```
Comment (aababilov) [inline core/src/main/java/org/mobilitydata/gtfsvalidator/table/GtfsTableContainerWithMultiColumnPrimaryKey.java:18]: Typo: MultiColumnKey

File: main/src/main/java/org/mobilitydata/gtfsvalidator/table/GtfsCalendarDateSchema.java [hunk 6]
```diff
@@ -29,10 +30,12 @@
 public interface GtfsCalendarDateSchema extends GtfsEntity {
   @FieldType(FieldTypeEnum.ID)
   @Required
+  @PrimaryKey
```
Comment (aababilov) [inline main/src/main/java/org/mobilitydata/gtfsvalidator/table/GtfsCalendarDateSchema.java:33]: Changes to this GtfsCalendarDateSchema and GtfsFrequencySchema make sense to me. However, they are adding extra validations and may potentially reject some feeds that are considered valid right now. Can we modify these files in a separate pull request?

**Review by aababilov — CHANGES_REQUESTED** [review]
I like this change but please see my comments.

File: processor/src/main/java/org/mobilitydata/gtfsvalidator/processor/EndRangeValidatorGenerator.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (bdferris-v2) [reply processor/src/main/java/org/mobilitydata/gtfsvalidator/processor/EndRangeValidatorGenerator.java:151]: It's only a problem for multi-column primary keys that must be discussed in notices that only support single columns.

File: main/src/main/java/org/mobilitydata/gtfsvalidator/validator/TranslationFieldAndReferenceValidator.java [hunk 7]
```diff
@@ -189,6 +200,52 @@ private static boolean isMissingOrUnexpectedField(
     return true;
   }
 
+  private Optional<?> resolveRecord(
```
[the remark this answers is not in the document]
Comment (bdferris-v2) [reply main/src/main/java/org/mobilitydata/gtfsvalidator/validator/TranslationFieldAndReferenceValidator.java:203]: So this gets into the extended discussion we had about key ordering in https://bit.ly/gtfs-validator-multi-column-primary-keys.  Namely, translations.txt effectively defines a primary key order to be used when performing entity resolution from the recordId + recordSubId fields.  Indeed, the first version of this pull request has a more generic method here, but it hinged on determining the implicit key ordering from the order of fields in the Java schema file.  Ultimately, I was worried that relying on that was shaky at best and since the concept of key ordering was translations.txt-specific, I made the call to move the code into the translations.txt-specific validator.

Now, I will definitely concede that the resulting code, while type-safe, is pretty ugly :)  It's kind of begging for code-generation, but I wasn't sure I wanted to apply that hammer again quite yet.

Namely, I think the arguments about key order are still there.  But maybe there is a middle path here?

Maybe we could define `enum TranslationRecordIdType { RECORD_ID, RECORD_SUB_ID }` and then add a field to PrimaryKey:

```
@interface PrimaryKey {
  ...
  TranslationRecordIdType translationRecordIdType() default RECORD_ID;
}
```

Basically, instead of relying on key-order, we are directly tying the @PrimaryKey definition to the semantics of `translations.txt`.  And because of the default value semantics, we'd really only need to define RECORD_SUB_ID overrides for a handful of fields on existing multi-column schemas.  Generating an appropriate generic lookup method from those annotations should be doable and it doesn't lock us into a more general argument about primary key ordering.

Anyone have thoughts?  @aababilov or @maximearmstrong

File: main/src/main/java/org/mobilitydata/gtfsvalidator/table/GtfsCalendarDateSchema.java (hunk 6 above)
Comment (bdferris-v2) [reply main/src/main/java/org/mobilitydata/gtfsvalidator/table/GtfsCalendarDateSchema.java:33]: There is going to be a tricky chicken-and-egg problem here.  Depending on what mechanism we come up for translations.txt validation support, we need to come up with something that can gracefully handle having the changes to calendar date and frequency split into a separate PR.  Let's see how the other conversation goes before making this change.

File: processor/src/main/java/org/mobilitydata/gtfsvalidator/processor/EndRangeValidatorGenerator.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (maximearmstrong) [reply processor/src/main/java/org/mobilitydata/gtfsvalidator/processor/EndRangeValidatorGenerator.java:151]: I understand. If you think this refactor would valuable in the long run, then we should plan it. I think it would be part v4. Though, I think your changes make sense for now. cc @isabelle-dr

File: main/src/main/java/org/mobilitydata/gtfsvalidator/validator/TranslationFieldAndReferenceValidator.java (hunk 7 above)
[the remark this answers is not in the document]
Comment (maximearmstrong) [reply main/src/main/java/org/mobilitydata/gtfsvalidator/validator/TranslationFieldAndReferenceValidator.java:203]: I would agree with that last idea - `enum TranslationRecordIdType` could be a good middle path. Do you think it should in this PR or another?

I kinda agree with both of you - a more generic validator would be good, but the solution brought by @bdferris-v2 makes sense for what we are trying to accomplish here. I would be okay with this being the initial solution and then we can consider a refactor using `enum TranslationRecordIdType`. Do you think it'd be a good plan?

[the remark this answers is not in the document]
Comment (bdferris-v2) [reply main/src/main/java/org/mobilitydata/gtfsvalidator/validator/TranslationFieldAndReferenceValidator.java:203]: I'd be willing to update this PR if you both thought this was a path worth taking (aka the `TranslationRecordIdType` proposal).

[the remark this answers is not in the document]
Comment (maximearmstrong) [reply main/src/main/java/org/mobilitydata/gtfsvalidator/validator/TranslationFieldAndReferenceValidator.java:203]: I'm fine with that. What about you @aababilov?

[the remark this answers is not in the document]
Comment (aababilov) [reply main/src/main/java/org/mobilitydata/gtfsvalidator/validator/TranslationFieldAndReferenceValidator.java:203]: I love this idea! Let's add TranslationRecordIdType in this PR.

File: main/src/main/java/org/mobilitydata/gtfsvalidator/table/GtfsCalendarDateSchema.java (hunk 6 above)
Comment (aababilov) [reply main/src/main/java/org/mobilitydata/gtfsvalidator/table/GtfsCalendarDateSchema.java:33]: I am fine with changing GtfsCalendarDateSchema and GtfsFrequencySchema here if we don't have any feeds at Google that break this requirement (from thousands of feeds, there is usually at least one that breaks any assumption we make. Kind of Hyrum's Law for GTFS).

**Review by aababilov — CHANGES_REQUESTED** [review]
Sorry for the delay! And thanks for the nice PR.

File: main/src/main/java/org/mobilitydata/gtfsvalidator/validator/TranslationFieldAndReferenceValidator.java (hunk 7 above)
[the remark this answers is not in the document]
Comment (bdferris-v2) [reply main/src/main/java/org/mobilitydata/gtfsvalidator/validator/TranslationFieldAndReferenceValidator.java]: Done

**Review by aababilov — APPROVED** [review]
Thank you!

File: core/src/main/java/org/mobilitydata/gtfsvalidator/annotation/PrimaryKey.java [hunk 8]
```diff
@@ -51,7 +55,20 @@
...
+public @interface PrimaryKey {
+
+  /**
+   * Typically used in a multi-column primary key situation. When true, if an index is generated for
+   * other primary key fields in the file, the associated list of values will be sorted by the
+   * sequence field. Classic examples include stop_times.txt, where the primary key is (trip_id,
+   * stop_sequence) and a byTripId(...) method returns a list of stop-times that are sorted by their
+   * `stop_sequence` value.
+   */
+  boolean isSequence() default false;
```
[the remark this answers is not in the document]
Comment (bdferris-v2) [reply core/src/main/java/org/mobilitydata/gtfsvalidator/annotation/PrimaryKey.java]: No, I don't think we do for now.  I added a compile-time check in `GtfsFileDescriptor` restricting it to one for now.

File: core/src/main/java/org/mobilitydata/gtfsvalidator/annotation/PrimaryKey.java [hunk 9]
```diff
@@ -41,7 +41,11 @@
  *   }
  * </pre>
  *
- * <p>This also adds a method to the container class to find an entity by its primary key.
+ * <p>The {@pcode PrimaryKey} annotation can be specified for multiple fields if the file has a
+ * multi-column key.
```
[the remark this answers is not in the document]
Comment (bdferris-v2) [reply core/src/main/java/org/mobilitydata/gtfsvalidator/annotation/PrimaryKey.java:45]: For now, I believe we have agreed to not make any claims about the order.

File: core/src/main/java/org/mobilitydata/gtfsvalidator/annotation/PrimaryKey.java [hunk 10]
```diff
@@ -51,7 +55,20 @@
  *       }
  *   }
  * </pre>
+ *
+ * <p>A lookup method will not automatically be generated for multi-field primary keys. Instead, use
+ * the {@code @Index} annotation to specify which individual fields should have lookup methods.
```
[the remark this answers is not in the document]
Comment (bdferris-v2) [reply core/src/main/java/org/mobilitydata/gtfsvalidator/annotation/PrimaryKey.java:60]: I don't believe so.  This PR shouldn't change any logic there.

File: core/src/main/java/org/mobilitydata/gtfsvalidator/table/GtfsTableContainer.java [hunk 11]
```diff
@@ -92,11 +86,10 @@ public boolean hasColumn(String columnName) {
    * Note that all keys must be passed as strings, so a key for {@code stop_times.txt} will look
    * like {"stop1", "0"}.
    *
-   * @param id primary key or first part of the composite key, if needed
-   * @param subId second part of the composite key, if needed
+   * @param ids primary key or first part of the composite key, if needed
```
[the remark this answers is not in the document]
Comment (bdferris-v2) [reply core/src/main/java/org/mobilitydata/gtfsvalidator/table/GtfsTableContainer.java]: No longer applies.

File: processor/src/main/java/org/mobilitydata/gtfsvalidator/processor/GtfsFileDescriptor.java [hunk 12]
```diff
@@ -92,14 +95,11 @@ public abstract static class Builder {
     public GtfsFileDescriptor build() {
       ImmutableMap.Builder<String, GtfsFieldDescriptor> fieldsMapBuilder = ImmutableMap.builder();
       ImmutableList.Builder<GtfsFieldDescriptor> indicesBuilder = ImmutableList.builder();
+      ImmutableList.Builder<GtfsFieldDescriptor> primaryKeysBuilder = ImmutableList.builder();
       for (GtfsFieldDescriptor field : fields()) {
```
[the remark this answers is not in the document]
Comment (bdferris-v2) [reply processor/src/main/java/org/mobilitydata/gtfsvalidator/processor/GtfsFileDescriptor.java:112]: Potentially?  Assuming the code analyzer processes fields in file-order.  But none of the downstream code depends on the order matching, so I'm not sure it's worth worrying about?

File: main/src/main/java/org/mobilitydata/gtfsvalidator/table/GtfsCalendarDateSchema.java (hunk 6 above)
Comment (bdferris-v2) [reply main/src/main/java/org/mobilitydata/gtfsvalidator/table/GtfsCalendarDateSchema.java:35]: I did a large-scale eval on production feeds and didn't find any new validation errors when run against the new @PrimaryKey additions.  I think this is good to go.

Comment (aababilov) [reply main/src/main/java/org/mobilitydata/gtfsvalidator/table/GtfsCalendarDateSchema.java:35]: Thanks for running the experiment!
